syntax = "proto2";

option cc_enable_arenas = true;

package c2018.score_subsystem;

enum ScoreGoal {
  IDLE_BOTTOM = 0;
  IDLE_STOW = 1;
  HEIGHT_0 = 2;
  HEIGHT_1 = 3;
  HEIGHT_2 = 4;
  PREP_SCORE_LOW = 5;
  PREP_SCORE_MID = 6;
  PREP_SCORE_MID_BACK = 7;
  PREP_SCORE_HIGH = 8;
  PREP_SCORE_HIGH_BACK = 9;
  SCORE_LOW = 10;
  SCORE_MID = 11;
  SCORE_MID_BACK = 12;
  SCORE_HIGH = 13;
  SCORE_HIGH_BACK = 14;
  INTAKE_MANUAL = 15;
  OUTTAKE_MANUAL = 16;
  IDLE_MANUAL = 17;
}

enum IntakeMode {
  IDLE = 0;
  INTAKE = 1;
  OUTTAKE = 2;
}

message ScoreSubsystemInput {
  required uint64 timestamp = 1;
  required double elevator_encoder = 2;
  required bool elevator_hall = 3;
  required double wrist_encoder = 4;
  required bool wrist_hall = 5;
  required double intake_current = 6;
  required bool has_cube = 7;
}

message ScoreSubsystemOutput {
  required uint64 timestamp = 1;
  required double elevator_voltage = 2;
  required double intake_voltage = 3;
  required double wrist_voltage = 4;
  required bool wrist_solenoid_open = 5;
  required bool wrist_solenoid_close = 6;
}

message ScoreSubsystemGoal {
  required uint64 timestamp = 1;
  optional double wrist_angle = 2;
  required double elevator_voltage = 3;
  required double roller_voltage = 4;
  required double wrist_voltage = 5;
  required bool claw_pinch = 6;
  required ScoreGoal score_goal = 7;
  required bool score_forward = 8;

}

message ScoreSubsystemStatus {
  required uint64 timestamp = 1;
  required double elevator_actual_height = 2;
  required double estimated_velocity = 3;
  required bool elevator_at_top = 4;
  required double elevator_voltage_error = 5;
  required double elevator_uncapped_voltage = 6;
  required double elevator_profiled_goal = 7;
  required double elevator_unprofiled_goal = 8;
  required bool elevator_encoder_fault_detected = 9;
  required bool elevator_calibrated = 11;
  required double wrist_angle = 12;
  required double wrist_velocity = 13;
  required bool wrist_calibrated = 14;
  required double wrist_profiled_goal = 15;
  required double wrist_unprofiled_goal = 16;
  required bool claw_pinch = 17;
  required bool has_cube = 18;
}
